From 11ddc22e7fd72b003cf3694468b6d9fc03de3993 Mon Sep 17 00:00:00 2001
From: Viorel Suman <viorel.suman@nxp.com>
Date: Fri, 23 Feb 2018 14:09:36 +0200
Subject: [PATCH 4099/5242] MLK-18691: ASoC: imx-xtor: let DAI calculate
 SYSCLK freq needed

commit  b8e022fb033edb4c4ccd38f8c858665138746c4c from
https://source.codeaurora.org/external/imx/linux-imx.git

Currently SAI interface sets MCLK rate if provisioned
externally from machine driver through set_sysclk. Let
SAI manage clock frequency without interfering from machine driver.
Aside of this use channels number to set TDM tx and rx masks.

Signed-off-by: Viorel Suman <viorel.suman@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 sound/soc/fsl/imx-xtor.c |    9 ++++-----
 1 file changed, 4 insertions(+), 5 deletions(-)

diff --git a/sound/soc/fsl/imx-xtor.c b/sound/soc/fsl/imx-xtor.c
index 0d6f92d..3b59553 100644
--- a/sound/soc/fsl/imx-xtor.c
+++ b/sound/soc/fsl/imx-xtor.c
@@ -78,9 +78,9 @@ static int imx_xtor_hw_params(struct snd_pcm_substream *substream,
 	bool tx = substream->stream == SNDRV_PCM_STREAM_PLAYBACK;
 	struct cpu_priv *cpu_priv = &data->cpu_priv;
 	struct device *dev = rtd->card->dev;
+	u32 channels = params_channels(params);
 	unsigned int fmt = SND_SOC_DAIFMT_I2S | SND_SOC_DAIFMT_NB_NF;
 	int ret, dir;
-	u32 freq;
 
 	/* For playback the XTOR is slave, and for record is master */
 	fmt |= tx ? SND_SOC_DAIFMT_CBS_CFS : SND_SOC_DAIFMT_CBM_CFM;
@@ -94,16 +94,15 @@ static int imx_xtor_hw_params(struct snd_pcm_substream *substream,
 	}
 
 	/* Specific configurations of DAIs starts from here */
-	freq = params_rate(params) * params_width(params) * cpu_priv->slots;
 	ret = snd_soc_dai_set_sysclk(rtd->cpu_dai, cpu_priv->sysclk_id[tx],
-				     freq, dir);
+				     0, dir);
 	if (ret) {
 		dev_err(dev, "failed to set cpu sysclk: %d\n", ret);
 		return ret;
 	}
 
-	ret = snd_soc_dai_set_tdm_slot(rtd->cpu_dai, 0x3, 0x3, cpu_priv->slots,
-					params_width(params));
+	ret = snd_soc_dai_set_tdm_slot(rtd->cpu_dai, BIT(channels) - 1,
+		BIT(channels) - 1, cpu_priv->slots, params_width(params));
 	if (ret) {
 		dev_err(dev, "failed to set cpu dai tdm slot: %d\n", ret);
 		return ret;
-- 
1.7.9.5

